File: Julia/Gemini/NBodySimulationKDop.jl
# Okay, let's implement the parallel kD-tree construction using the ParallelKDTrees.jl package. This is a significant optimization that can speed up the simulation, especially for a large number of bodies.
# First, ensure you have the ParallelKDTrees package installed:
# Julia

# using Pkg
# Pkg.add("ParallelKDTrees")
# Now, here's the modified code incorporating parallel kD-tree construction. Note that the calculate_force_kd_tree! function needs to be adapted to work with the KDTree structure from NearestNeighbors (which ParallelKDTrees builds upon).
#To Run:
#
#Save the code as a .jl file (e.g., nbody_parallel_kdtree.jl).
#Ensure you have NearestNeighbors and ParallelKDTrees installed.
#Run Julia with multiple threads: julia -t <num_threads> nbody_parallel_kdtree.jl

using LinearAlgebra
using Base.Threads
using NearestNeighbors
using ParallelKDTrees

const GRAVITATIONAL_CONSTANT = 6.67430e-11
const THETA = 0.3

mutable struct Body
    mass::Float64
    position::Vec3{Float64}
    velocity::Vec3{Float64}
    force::Vec3{Float64} # To store calculated force
end

struct System
    bodies::Vector{Body}
end

distance_squared(p1::Vec3{Float64}, p2::Vec3{Float64}) = sum((p1 - p2).^2)
distance(p1::Vec3{Float64}, p2::Vec3{Float64}) = sqrt(distance_squared(p1, p2))

function calculate_force_element!(body_i::Body, mass_element_cm::Vec3{Float64}, mass_element_mass::Float64)
    r_sq = distance_squared(body_i.position, mass_element_cm)
    if r_sq > 1e-9
        r = sqrt(r_sq)
        magnitude = (GRAVITATIONAL_CONSTANT * body_i.mass * mass_element_mass) / r_sq
        body_i.force .+= magnitude * (mass_element_cm - body_i.position) / r
    end
end

function calculate_total_energy(system::System)
    kinetic_energy = Threads.Atomic{Float64}(0.0)
    potential_energy = Threads.Atomic{Float64}(0.0)
    num_bodies = length(system.bodies)

    @threads for i in 1:num_bodies
        body_i = system.bodies[i]
        v_sq = sum(body_i.velocity.^2)
        Threads.atomic_add!(kinetic_energy, 0.5 * body_i.mass * v_sq)

        for j in i+1:num_bodies
            body_j = system.bodies[j]
            r = distance(body_i.position, body_j.position)
            Threads.atomic_add!(potential_energy, -(GRAVITATIONAL_CONSTANT * body_i.mass * body_j.mass) / r)
        end
    end
    return Threads.load(kinetic_energy) + Threads.load(potential_energy)
end

function initialize_circular_orbits(num_orbiting::Int, central_mass::Float64, orbit_radius::Float64, orbiting_mass::Float64)
    bodies = Vector{Body}()
    # Initialize the central body
    push!(bodies, Body(central_mass, zeros(Vec3{Float64}), zeros(Vec3{Float64}), zeros(Vec3{Float64})))

    # Initialize the orbiting bodies
    for i in 0:(num_orbiting - 1)
        angle = 2 * pi * i / num_orbiting
        x = orbit_radius * cos(angle)
        y = orbit_radius * sin(angle)
        z = 0.0 # Orbiting in the xy-plane
        position = Vec3(x, y, z)

        # Calculate the orbital velocity for a circular orbit
        orbital_speed = sqrt(GRAVITATIONAL_CONSTANT * central_mass / orbit_radius)
        vx = -orbital_speed * sin(angle)
        vy = orbital_speed * cos(angle)
        vz = 0.0
        velocity = Vec3(vx, vy, vz)

        push!(bodies, Body(orbiting_mass, position, velocity, zeros(Vec3{Float64})))
    end
    return System(bodies)
end

function build_kd_tree(system::System)
    num_bodies = length(system.bodies)
    if num_bodies == 0
        return nothing
    end
    positions = [b.position for b in system.bodies]
    return KDTree(positions, leafsize=10) # Build in parallel by default in ParallelKDTrees
end

function calculate_force_kd_tree!(body::Body, tree::KDTree, system::System)
    body_index = findfirst(b -> b === body, system.bodies)
    if isnothing(body_index)
        return
    end
    current_pos = body.position
    force = zeros(Vec3{Float64})

    function barnes_hut!(node_indices::Vector{Int}, bounding_box::HyperRectangle{3, Float64})
        if isempty(node_indices)
            return
        elseif length(node_indices) == 1
            other_body_index = node_indices[1]
            if other_body_index != body_index
                other_body = system.bodies[other_body_index]
                calculate_force_element!(body, other_body.position, other_body.mass) # Treat as point mass
            end
        else
            s = maximum(bounding_box.max - bounding_box.min)
            center_of_mass = sum(system.bodies[i].mass * system.bodies[i].position for i in node_indices) / sum(system.bodies[i].mass for i in node_indices)
            d = distance(current_pos, center_of_mass)

            if d == 0.0 || s / d < THETA
                total_mass = sum(system.bodies[i].mass for i in node_indices)
                calculate_force_element!(body, center_of_mass, total_mass) # Treat as pseudo-particle
            else
                # Subdivide the box and recurse
                mid = (bounding_box.min + bounding_box.max) / 2
                for i in 0:7 # 2^3 octants in 3D
                    octant_indices = Int[]
                    octant_min = zeros(Vec3{Float64})
                    octant_max = zeros(Vec3{Float64})
                    for dim in 1:3
                        if (i >> (dim - 1)) & 1 == 0
                            octant_min[dim] = bounding_box.min[dim]
                            octant_max[dim] = mid[dim]
                        else
                            octant_min[dim] = mid[dim]
                            octant_max[dim] = bounding_box.max[dim]
                        end
                    end
                    octant_box = HyperRectangle(octant_min, octant_max)
                    indices_in_octant = filter(idx -> is_point_in_bounds(system.bodies[idx].position, octant_box), node_indices)
                    barnes_hut!(indices_in_octant, octant_box)
                end
            end
        end
    end

    root_indices = 1:length(system.bodies)
    min_coords = minimum(b.position for b in system.bodies)
    max_coords = maximum(b.position for b in system.bodies)
    root_bounds = HyperRectangle(min_coords, max_coords)

    barnes_hut!(root_indices, root_bounds)
    body.force .= force # Assign the accumulated force
end

function is_point_in_bounds(point::Vec3{Float64}, bounds::HyperRectangle{3, Float64})
    return all(bounds.min .<= point .<= bounds.max)
end

function kick_step!(system::System, dt::Float64)
    num_bodies = length(system.bodies)
    tree = build_kd_tree(system)

    # Reset forces
    @threads for i in 1:num_bodies
        system.bodies[i].force .= 0.0
    end

    # Calculate forces using kD-tree in parallel
    @threads for i in 1:num_bodies
        calculate_force_kd_tree!(system.bodies[i], tree, system)
    end

    # Update velocities (kick)
    @threads for i in 1:num_bodies
        body = system.bodies[i]
        body.velocity .+= (body.force / body.mass) * (dt / 2)
    end
end

function drift_step!(system::System, dt::Float64)
    @threads for body in system.bodies
        body.position .+= body.velocity * dt
    end
end

function first_order_kick_step!(system::System, dt::Float64)
    kick_step!(system, dt)
    drift_step!(system, dt)
    kick_step!(system, dt)
end

function main()
    num_orbiting_bodies = 1_000_000
    central_mass = 1.989e30    # Mass of the Sun (kg)
    orbit_radius = 1.496e11    # 1 AU (m)
    orbiting_mass = 5.972e24   # Mass of the Earth (kg)
    num_steps = 1000
    time_step = 3600.0 * 24.0 * 7.0 # 1 week in seconds

    # Initialize the system
    global sys = initialize_circular_orbits(num_orbiting_bodies, central_mass, orbit_radius, orbiting_mass)
    initial_system = deepcopy(sys)

    println("Initial number of bodies: $(length(sys.bodies))")

    # Calculate initial energy
    initial_energy = calculate_total_energy(initial_system)
    println("Initial total energy: $(initial_energy) J")

    println("Running simulation for $num_steps steps...")
    start_time = time()
    for step in 1:num_steps
        first_order_kick_step!(sys, time_step)
        if step % 100 == 0
            println("Step $step completed.")
        end
    end
    end_time = time()
    elapsed_time = end_time - start_time
    println("Simulation finished in $(elapsed_time) seconds.")

    # Calculate final energy
    final_energy = calculate_total_energy(sys)
    println("Final total energy: $(final_energy) J")

    # Calculate the energy difference
    energy_difference = abs(final_energy - initial_energy)
    relative_energy_difference = energy_difference / abs(initial_energy)
    println("Absolute energy difference: $(energy_difference) J")
    println("Relative energy difference: $(relative_energy_difference)")
end

main()

File: Go/Gemini/NBodySimulationKDop.go
// Okay, let's implement the kickStepKDTreeOptimized function with chunking for the parallel force calculation. This is a relatively straightforward optimization to try and can potentially improve performance by reducing the overhead of launching a goroutine for each individual body.
/* Changes Made:

kickStepKDTree Modified:
The function now takes an additional chunkSize parameter.
The loop that launches goroutines now iterates with a step of chunkSize.
Each goroutine processes a slice of the bodies and forces arrays.
firstOrderKickStepKDTree Modified:
It now calls kickStepKDTree with the chunkSize.
main Function Modified:
A chunkSize variable is introduced (set to 1000 as an example). You should experiment with different values for this based on your system's core count and performance.
firstOrderKickStepKDTree is called with the chunkSize.
How Chunking Can Help:

Reduced Goroutine Overhead: Launching a single goroutine to process a batch of bodies is generally more efficient than launching a goroutine for each individual body, especially when the number of bodies is very large. The overhead of goroutine creation and scheduling can become significant.
Improved Cache Locality: Processing a contiguous chunk of bodies might lead to better cache utilization as the data for these bodies is likely to be closer in memory.
**Next Steps for Further Optimization (Beyond this immediate change):
*/

package main

import (
	"fmt"
	"math"
	"math/rand"
	"sync"
	"time"

	"github.com/rudty/go-parallel"
)

// Body represents a celestial body with mass, position, and velocity.
type Body struct {
	Mass     float64
	Position [3]float64
	Velocity [3]float64
}

// System represents a collection of celestial bodies.
type System struct {
	Bodies []Body
}

// KDNode represents a node in the kD-tree.
type KDNode struct {
	bodyIndex    int // Index of the body in this leaf node (-1 if internal)
	centerOfMass [3]float64
	totalMass    float64
	min          [3]float64 // Minimum bounds of the node
	max          [3]float64 // Maximum bounds of the node
	left         *KDNode
	right        *KDNode
}

// GravitationalConstant is the universal gravitational constant.
const GravitationalConstant = 6.67430e-11
const theta = 0.3 // Barnes-Hut opening angle

// distanceSquared calculates the squared Euclidean distance between two points.
func distanceSquaredPoints(p1, p2 [3]float64) float64 {
	dx := p1[0] - p2[0]
	dy := p1[1] - p2[1]
	dz := p1[2] - p2[2]
	return dx*dx + dy*dy + dz*dz
}

// distancePoints calculates the Euclidean distance between two points.
func distancePoints(p1, p2 [3]float64) float64 {
	return math.Sqrt(distanceSquaredPoints(p1, p2))
}

// calculateForceElement calculates the force on a body due to a mass element (node or body).
func calculateForceElement(body *Body, massElementCM [3]float64, massElementMass float64, force *[3]float64) {
	rSq := distanceSquaredPoints(body.Position, massElementCM)
	if rSq > 1e-9 {
		r := math.Sqrt(rSq)
		magnitude := (GravitationalConstant * body.Mass * massElementMass) / rSq
		force[0] += magnitude * (massElementCM[0] - body.Position[0]) / r
		force[1] += magnitude * (massElementCM[1] - body.Position[1]) / r
		force[2] += magnitude * (massElementCM[2] - body.Position[2]) / r
	}
}

// calculateTotalEnergy calculates the total energy of the system.
func calculateTotalEnergy(s *System) float64 {
	kineticEnergy := 0.0
	potentialEnergy := 0.0
	numBodies := len(s.Bodies)

	var wg sync.WaitGroup
	var kineticMutex sync.Mutex
	var potentialMutex sync.Mutex

	wg.Add(numBodies)
	for i := 0; i < numBodies; i++ {
		go func(i int) {
			defer wg.Done()
			vSq := s.Bodies[i].Velocity[0]*s.Bodies[i].Velocity[0] +
				s.Bodies[i].Velocity[1]*s.Bodies[i].Velocity[1] +
				s.Bodies[i].Velocity[2]*s.Bodies[i].Velocity[2]
			kineticMutex.Lock()
			kineticEnergy += 0.5 * s.Bodies[i].Mass * vSq
			kineticMutex.Unlock()

			for j := i + 1; j < numBodies; j++ {
				r := math.Sqrt(distanceSquared(&s.Bodies[i], &s.Bodies[j]))
				potentialMutex.Lock()
				potentialEnergy -= (GravitationalConstant * s.Bodies[i].Mass * s.Bodies[j].Mass) / r
				potentialMutex.Unlock()
			}
		}(i)
	}
	wg.Wait()

	return kineticEnergy + potentialEnergy
}

// initializeCircularOrbits initializes a system with a central body and orbiting smaller bodies.
func initializeCircularOrbits(numOrbiting int, centralMass, orbitRadius, orbitingMass float64) *System {
	system := &System{
		Bodies: make([]Body, numOrbiting+1),
	}

	// Initialize the central body
	system.Bodies[0] = Body{
		Mass:     centralMass,
		Position: [3]float64{0, 0, 0},
		Velocity: [3]float64{0, 0, 0},
	}

	// Initialize the orbiting bodies
	for i := 0; i < numOrbiting; i++ {
		angle := 2 * math.Pi * float64(i) / float64(numOrbiting)
		system.Bodies[i+1] = Body{
			Mass: orbitingMass,
			Position: [3]float64{
				orbitRadius * math.Cos(angle),
				orbitRadius * math.Sin(angle),
				0,
			},
		}

		// Calculate the orbital velocity for a circular orbit
		orbitalSpeed := math.Sqrt(GravitationalConstant * centralMass / orbitRadius)
		system.Bodies[i+1].Velocity = [3]float64{
			-orbitalSpeed * math.Sin(angle),
			orbitalSpeed * math.Cos(angle),
			0,
		}
	}

	return system
}

// calculateBounds calculates the bounding box of a set of bodies.
func calculateBounds(bodies []Body) ([3]float64, [3]float64) {
	if len(bodies) == 0 {
		return [3]float64{}, [3]float64{}
	}
	min := bodies[0].Position
	max := bodies[0].Position
	for _, b := range bodies {
		for i := 0; i < 3; i++ {
			if b.Position[i] < min[i] {
				min[i] = b.Position[i]
			}
			if b.Position[i] > max[i] {
				max[i] = b.Position[i]
			}
		}
	}
	return min, max
}

// newKDNode creates a new kD-tree node.
func newKDNode(bodyIndex int) *KDNode {
	return &KDNode{
		bodyIndex: bodyIndex,
		totalMass: 0,
		left:      nil,
		right:     nil,
	}
}

// buildKDTreeRecursive recursively builds the kD-tree.
func buildKDTreeRecursive(bodies []Body, indices []int, minBound, maxBound [3]float64, depth int) *KDNode {
	numIndices := len(indices)
	if numIndices == 0 {
		return nil
	}

	if numIndices == 1 {
		node := newKDNode(indices[0])
		node.totalMass = bodies[indices[0]].Mass
		node.centerOfMass = bodies[indices[0]].Position
		node.min = minBound
		node.max = maxBound
		return node
	}

	node := newKDNode(-1) // Internal node
	node.min = minBound
	node.max = maxBound
	node.totalMass = 0
	centerOfMass := [3]float64{0, 0, 0}
	for _, index := range indices {
		node.totalMass += bodies[index].Mass
		for i := 0; i < 3; i++ {
			centerOfMass[i] += bodies[index].Mass * bodies[index].Position[i]
		}
	}
	if node.totalMass > 0 {
		for i := 0; i < 3; i++ {
			node.centerOfMass[i] = centerOfMass[i] / node.totalMass
		}
	}

	splitDim := depth % 3
	median := (minBound[splitDim] + maxBound[splitDim]) / 2

	leftIndices := make([]int, 0)
	rightIndices := make([]int, 0)

	for _, index := range indices {
		if bodies[index].Position[splitDim] <= median {
			leftIndices = append(leftIndices, index)
		} else {
			rightIndices = append(rightIndices, index)
		}
	}

	leftMinBound := minBound
	leftMaxBound := maxBound
	leftMaxBound[splitDim] = median

	rightMinBound := minBound
	rightMaxBound := maxBound
	rightMinBound[splitDim] = median

	node.left = buildKDTreeRecursive(bodies, leftIndices, leftMinBound, leftMaxBound, depth+1)
	node.right = buildKDTreeRecursive(bodies, rightIndices, rightMinBound, rightMaxBound, depth+1)

	return node
}

// buildKDTree builds the kD-tree for the given system.
func buildKDTree(s *System) *KDNode {
	indices := make([]int, len(s.Bodies))
	for i := range s.Bodies {
		indices[i] = i
	}
	minBound, maxBound := calculateBounds(s.Bodies)
	return buildKDTreeRecursive(s.Bodies, indices, minBound, maxBound, 0)
}

// calculateForceKDTreeRecursive recursively calculates the force on a body using the kD-tree.
func calculateForceKDTreeRecursive(body *Body, node *KDNode, s *System, force *[3]float64) {
	if node == nil {
		return
	}

	if node.bodyIndex != -1 { // Leaf node
		if node.bodyIndex != (body - &s.Bodies[0]) {
			calculateForceElement(body, node.centerOfMass, node.totalMass, force)
		}
		return
	}

	// Internal node
	sNode := [3]float64{}
	for i := 0; i < 3; i++ {
		sNode[i] = node.max[i] - node.min[i]
	}
	d := distancePoints(body.Position, node.centerOfMass)

	if d == 0 || (sNode[0]/d < theta && sNode[1]/d < theta && sNode[2]/d < theta) {
		calculateForceElement(body, node.centerOfMass, node.totalMass, force)
	} else {
		calculateForceKDTreeRecursive(body, node.left, s, force)
		calculateForceKDTreeRecursive(body, node.right, s, force)
	}
}

// calculateForceKDTree calculates the force on a body using the kD-tree.
func calculateForceKDTree(body *Body, root *KDNode, s *System, force *[3]float64) {
	force[0] = 0
	force[1] = 0
	force[2] = 0
	calculateForceKDTreeRecursive(body, root, s, force)
}

// kickStepKDTree performs the "kick" step using the kD-tree for force calculation with chunking.
func kickStepKDTree(s *System, dt float64, chunkSize int) {
	numBodies := len(s.Bodies)
	forces := make([][3]float64, numBodies)
	root := buildKDTree(s)

	// pool := parallel.NewGroup()
	// pool.SetMaxWorkers(runtime.NumCPU())

	for i := 0; i < numBodies; i += chunkSize {
		start := i
		end := i + chunkSize
		if end > numBodies {
			end = numBodies
		}

		// pool.Go(func() error {
		parallel.For(0, numBodies, func(i int) {
			for j := start; j < end; j++ {
				calculateForceKDTree(&s.Bodies[j], root, s, &forces[j])
			}
			// return nil
		})
	}
	// pool.Wait()

	// Update velocities
	for i := 0; i < numBodies; i++ {
		s.Bodies[i].Velocity[0] += (forces[i][0] / s.Bodies[i].Mass) * dt * 0.5
		s.Bodies[i].Velocity[1] += (forces[i][1] / s.Bodies[i].Mass) * dt * 0.5
		s.Bodies[i].Velocity[2] += (forces[i][2] / s.Bodies[i].Mass) * dt * 0.5
	}
}

// driftStep performs the "drift" part of the first-order kick-step method (parallelized).
func driftStep(s *System, dt float64) {
	numBodies := len(s.Bodies)
	// pool := parallel.NewGroup()
	// pool.SetMaxWorkers(runtime.NumCPU())

	for i := 0; i < numBodies; i++ {
		// i := i
		// pool.Go(func() error {
		parallel.For(0, numBodies, func(i int) {
			s.Bodies[i].Position[0] += s.Bodies[i].Velocity[0] * dt
			s.Bodies[i].Position[1] += s.Bodies[i].Velocity[1] * dt
			s.Bodies[i].Position[2] += s.Bodies[i].Velocity[2] * dt
			// return nil
		})
	}
	// pool.Wait()
}

// firstOrderKickStepKDTree performs one full first-order kick-step using the kD-tree with chunking.
func firstOrderKickStepKDTree(s *System, dt float64, chunkSize int) {
	kickStepKDTree(s, dt, chunkSize)
	driftStep(s, dt)
	kickStepKDTree(s, dt, chunkSize)
}

func main() {
	numOrbitingBodies := 100000
	centralMass := 1.989e30  // Mass of the Sun (kg)
	orbitRadius := 1.496e11  // 1 AU (m)
	orbitingMass := 5.972e24 // Mass of the Earth (kg)
	numSteps := 10
	timeStep := 3600.0 * 24.0 * 7.0 // 1 week in seconds
	chunkSize := 1000               // Experiment with different chunk sizes

	rand.Seed(time.Now().UnixNano())

	// Initialize the system
	system := initializeCircularOrbits(numOrbitingBodies, centralMass, orbitRadius, orbitingMass)
	fmt.Printf("Initial number of bodies: %d\n", len(system.Bodies))

	// Calculate initial energy
	initialEnergy := calculateTotalEnergy(system)
	fmt.Printf("Initial total energy: %e J\n", initialEnergy)

	// Run the simulation
	fmt.Printf("Running simulation for %d steps...\n", numSteps)
	startTime := time.Now()
	for step := 0; step < numSteps; step++ {
		firstOrderKickStepKDTree(system, timeStep, chunkSize)
		if (step+1)%100 == 0 {
			fmt.Printf("Step %d completed.\n", step+1)
		}
	}
	elapsedTime := time.Since(startTime)
	fmt.Printf("Simulation finished in %s.\n", elapsedTime)

	// Calculate final energy
	finalEnergy := calculateTotalEnergy(system)
	fmt.Printf("Final total energy: %e J\n", finalEnergy)

	// Calculate the energy difference
	energyDifference := math.Abs(finalEnergy - initialEnergy)
	relativeEnergyDifference := energyDifference / math.Abs(initialEnergy)
	fmt.Printf("Absolute energy difference: %e J\n", energyDifference)
	fmt.Printf("Relative energy difference: %e\n", relativeEnergyDifference)
}
